scnlib fixes/changes/updates needed to be integrated in VTK library

spyridon97 opened the pull request:
This PR includes all the necessary changes to integrate scnlib in VTK/ParaView as a third-party library and be used extensively for its reading/scanning capabilities.

The summary of the changes is:
1. Export functions implemented in impl.cpp
2. Use SCN_NODISCARD everywhere
3. fast_float: support version 7+
4. gcc: fix assert usage in constexpr function
5. gcc: fix template deduction warning
6. basic_scan_file_buffer: Remove current_view size 1 assumption
7. arg_reader: Fix errors when reading last part of contiguous segments from file
8. win32/default: Enforce only 1 successfully putback guarantee

spyridon97 committed:
Export functions implemented in impl.cpp

spyridon97 committed:
Use SCN_NODISCARD everywhere

spyridon97 committed:
fast_float: support version 7+

chars_format became an enum class

spyridon97 committed:
gcc: fix assert usage in constexpr function

spyridon97 committed:
gcc: fix template deduction warning

warning: 'basic_string_view' may not intend to support class template argument deduction [-Wctad-maybe-unsupported]

spyridon97 commented on CMakeLists.txt:
```diff
@@ -70,6 +72,14 @@ target_link_libraries(scn PRIVATE
 )
 set_library_flags(scn)
 
+include(GenerateExportHeader)
+generate_export_header(scn
+  EXPORT_MACRO_NAME SCN_EXPORT
+  EXPORT_FILE_NAME include/scn/scn_export.h)
```
would that affect other functions/classes that don't thave the SCN_EXPORT symbol?

mathstuf replied:
You can set it as a direct target property, so that it doesn't. Using the property initialization variable would affect any other target in that variable's purview as well.

spyridon97 committed:
fast_float: disable long double tests

long double type is not supported by fast_float.

As of fast_float 7.0.0 compiling with fast_float leads
to a compiler error, rather than selecting the wrong
overload.

spyridon97 commented:
@eliaskosunen we would appreciate if you could review this PR.

spyridon97 committed:
basic_scan_file_buffer: Remove current_view size 1 assumption

basic_scan_file_buffer<FileInterface>::sync was assuming that when buffering is off,
the current_view's size will always be 1, but fill can return an empty current_view.

spyridon97 committed:
arg_reader: Fix errors when reading last part of contiguous segments from file

The contiguous range might fail because it does not have all
the characters, so we fall back and scan the range as not contiguous

A basic_scan_file_buffer, when it has buffering, it can become
discontiguous while moving forward when executing
ranges::next(range.begin(), ranges::distance(crange.begin(), it)).
E.g. the contiguous buffer might have 4.1234, but the full number is 4.12345
For that reason, we fall back and scan the range as not contiguous

Also, it might not read all the digits because it does not have them all
E.g. if it has 1.1234e, but it needed 1.1234e-01, it will disregard the e.
For that reason, we fall back and scan the range as not contiguous

spyridon97 committed:
win32/default: Enforce only 1 successfully putback guarantee

https://en.cppreference.com/w/c/io/ungetc states that only 1 byte
is guaranteed to be put back. Even if linux/mac support 4KB,
windows and other platforms are very strict with 1 character,
and they can fail if you try 2-3-4 ungetc.

To that extent, no more characters than needed are ever consumed.

Also, when parsing from a file, if whitespace has been consumed,
it remains consumed, it's not put back, similar to fscanf.
https://pubs.opengroup.org/onlinepubs/000095399/functions/fscanf.html

spyridon97 commented:
@eliaskosunen another friendly ping.

spyridon97 commented:
@eliaskosunen another friendly ping.

spyridon97 commented:
@eliaskosunen another friendly ping.
